src/runtime/ocl/ocl_hardware_manager.cpp

```diff
@@ -55,6 +56,97 @@ ResultT info_query(const cl::Device& dev) {
   return r;
 }
 
+bool parse_ocl_version_string(const std::string& s, int& major_version_out) {
+  const std::string identifier = "OpenCL ";
+  const auto pos = s.find(identifier);
+  if(pos != 0)
+    return false;
+  std::string version_substring = s.substr(identifier.length());
+  auto dot_pos = version_substring.find(".");
+  if(dot_pos != std::string::npos) {
+    major_version_out = std::stoi(version_substring.substr(0, dot_pos));
+    return true;
+  }
+
+  return false;
+}
+
+bool should_include_platform(const std::string& platform_name, const cl::Platform& p) {
+  const bool show_all_devices = application::get_settings().get<setting::ocl_show_all_devices>();
+  if(show_all_devices)
+    return true;
+
+  std::string ocl_version;
+  cl_int err = p.getInfo(CL_PLATFORM_VERSION, &ocl_version);
+  if (err != CL_SUCCESS) {
+    print_warning(__hipsycl_here(),
+                  error_info{"ocl_hardware_manager: Could not retrieve OpenCL "
+                             "version for platform " +
+                                 platform_name,
+                             error_code{"CL", err}});
+    return false;
+  } else {
+    int ocl_version_major = 0;
+    if (!parse_ocl_version_string(ocl_version, ocl_version_major)) {
+      HIPSYCL_DEBUG_WARNING
+          << "ocl_hardware_manager: Could not parse OpenCL version string "
+              "of platform '"
+          << platform_name
+          << "'; hiding platform. Set ACPP_RT_OCL_SHOW_ALL_DEVICES=1 "
+              "to override. (OpenCL version string was: '"
+          << ocl_version << "')" << std::endl;
+      return false;
+    } else {
+      if (ocl_version_major < 3) {
+        HIPSYCL_DEBUG_WARNING
```

Comment:
I suggest offering a simple extension to check whether any devices were hidden (`bool sycl::backend::ext_adaptivecpp_has_hidden_devices()`?), so that `acpp-info` can display a warning in this case ("Note: some incompatible OpenCL devices hidden; set ACPP_RT_OCL_SHOW_ALL_DEVICES=1 to show them"), while the default behavior does not raise any repeated, unactionable, and usually irrelevant warnings.

Reply:
This proposed extension can also take into account whether any devices were hidden by `ACPP_VISIBILITY_MASK` (or even by backend-specific means like `(CUDA|ROCR)_VISIBLE_DEVICES`).

Such messages are useless when setting the filters inline (`ACPP_VISIBILITY_MASK="omp;ocl" install/bin/acpp-info`), but SLURM or various scripts might set such variables unbeknownst to user. It's not polite to override them, but it might be helpful to know of them.

Reply:
>  All AMD and Intel users will have at such a non-compatible platform in their default install, so such situation is hardly "unexpected" (unlike the "Could not retrieve/parse OpenCL version" messages above).

I don't have AMD OpenCL in my ROCm, so it depends on which packages you've actually installed. But to the point: Isn't it a matter of perspective what is "unexpected"?
I think you are probably thinking of the unaware end user, who does not know the difference between OpenCL and CUDA and then might be confused by the warning. I see this point.
But I think there could also be a class of users where silently hiding devices might be the unexpected thing. E.g. a user tries to get AdaptiveCpp to work on OpenCL devices that we are not familiar with (e.g. on mobile platforms). Or they are from an OpenCL vendor, and are looking to extent hardware support to SYCL. In that case, they might expect that their expected device shows up, and then it can be very helpful to be warned that it won't due to reasons XYZ.

Perhaps we need to distinguish between "commonly known OpenCL implementation with commonly known issues and alternative backend available" and the other case of an unknown OpenCL implementation?

Otherwise yeah, limiting the warning to `acpp-info` could be an option, if users are aware enough to try it. But even in that case it might be difficult to forward the reason for hiding the device through such a query. I'm not sure we can easily do that at the moment.

Reply:
>  But I think there could also be a class of users where silently hiding devices might be the unexpected thing. E.g. a user tries to get AdaptiveCpp to work on OpenCL devices that we are not familiar with (e.g. on mobile platforms). Or they are from an OpenCL vendor, and are looking to extent hardware support to SYCL. In that case, they might expect that their expected device shows up, and then it can be very helpful to be warned that it won't due to reasons XYZ.

I see the utility in this case, but such "advanced" users are (a) less numerous than the "casual" users (hopefully, as SYCL get more popular, much less numerous :slightly_smiling_face:); (b) likely will quickly try setting a higher `HIPSYCL_DEBUG_LEVEL`, where the message will be printed.

> Perhaps we need to distinguish between "commonly known OpenCL implementation with commonly known issues and alternative backend available" and the other case of an unknown OpenCL implementation?

Agree, that's perhaps the best solution. Rather straightforward and strikes the balance between the two groups.

Reply:
I have changed it now such that it
* first checks the platform name against commonly known problematic OpenCL implementations (NVIDIA, AMD, Intel FPGA Emulator). If there is a match, it rejects the platform with an `HIPSYCL_DEBUG_INFO` message.
* Otherwise it proceeds to check the platform and device capabilities as before. If it finds an issue there, it emits a warning.